### ISSUE TITLE
ReadTheDocs autodoc requires `reentry scan`

### ISSUE THREAD

**Issue by chrisjsewell** [aiidateam/aiida-core#4200]
Currently RTD has a bunch of autodoc errors, which cause the affected classes to be missing from the API documentation:
https://readthedocs.org/projects/aiida-core/builds/11321660/

Hopefully running `reentry scan` (somewhere?) should fix this

**Comment by greschd** [comment]
In some of my plugins, I use the following code in ``docs/source/conf.py``:

```python
import reentry
reentry.manager.scan()
```

**Comment by chrisjsewell** [comment]
cheers yeh that is what I was just adding now in a PR

**Comment by chrisjsewell** [comment]
err damn it, its still broken 🤷 https://readthedocs.org/projects/aiida-core/builds/11336882/

**Comment by chrisjsewell** [comment]
maybe something to do with how aiida is added to the path: https://github.com/aiidateam/aiida-core/blob/bb2098e7035eda1526960897dec02fbd0a1a6a14/docs/source/conf.py#L24-L25

which is incompatible with how reentry works? @ltalirz any ideas?

**Comment by sphuber** [comment]
Note that you can push a potential fix to a branch on `aiidateam` and then enable that branch on RTD to do a build. This way we can test if the fix works before merging

**Comment by sphuber** [comment]
> maybe something to do with how aiida is added to the path:
> 
> https://github.com/aiidateam/aiida-core/blob/bb2098e7035eda1526960897dec02fbd0a1a6a14/docs/source/conf.py#L24-L25
> 
> which is incompatible with how reentry works? @ltalirz any ideas?

This is not necessarily incompatible, but if we don't install the `aiida-core` package then the entry points won't be either. So the package must no longer be installed then? Sure that was the case at some point

**Comment by chrisjsewell** [comment]
yeh I was naively expecting to my "fix" to work 🤦

**Comment by sphuber** [comment]
It seems indeed that on RTD it is not actually installing the package, so of course the entry points won't be registered. I also notice that the build is marked as finished, despite the warnings. I guess they don't run with the nitpick option on. Think it is reasonable to think then that it has been broken for a long time and not necessarily since the docs-revamp and we simply never noticed?

**Comment by chrisjsewell** [comment]
I think so: as mentioned in #4104 you can set RTD to fail straight away on warnings: https://docs.readthedocs.io/en/latest/config-file/v2.html#sphinx-fail-on-warning

**Comment by sphuber** [comment]
Think [this branch](https://github.com/aiidateam/aiida-core/tree/fix/4200/read-the-docs) may do the trick. I triggered [a build](https://readthedocs.org/projects/aiida-core/builds/11337187/) to verify.

**Comment by greschd** [comment]
> 
> It seems indeed that on RTD it is not actually installing the package, so of course the entry points won't be registered. I also notice that the build is marked as finished, despite the warnings. I guess they don't run with the nitpick option on. Think it is reasonable to think then that it has been broken for a long time and not necessarily since the docs-revamp and we simply never noticed?

You can make readthedocs install the package.. either by fiddling with the online interface, or using a `.readthedocs.yaml` as e.g. [in aiida-wannier90](https://github.com/aiidateam/aiida-wannier90/blob/develop/.readthedocs.yml#L16).

**Comment by chrisjsewell** [comment]
oh yeh I didn't notice that wasn't set up before, who added the empty `.readthedocs.yaml` lol

**Comment by sphuber** [comment]
> **EDIT:** Just noticed that's what your branch is doing sweat_smile

Yeah, except for some reason it didn't seem to have picked up my changes. I am looking at the logs and it didn't install `aiida-core[docs]`. Not sure why. Cannot find a reference where the `.readthedocs.yaml` is read in

**Comment by greschd** [comment]
Hmm.. maybe because it's missing the `version` specifier? This is the doc for the YAML file: https://docs.readthedocs.io/en/stable/config-file/v2.html

**Comment by greschd** [comment]
Yeah, looking at this.. the current file seems to be in `v1` format, that didn't support most of these options.

**Comment by chrisjsewell** [comment]
maybe get rid of that `formats: []` and add `version: 2`: like here: https://github.com/executablebooks/MyST-Parser/blob/master/.readthedocs.yml

**Comment by greschd** [comment]
Removing `formats` shouldn't be necessary - it's the default value anyway, but since there's a long comment explaining why it's empty maybe explicit is better here.

**Comment by chrisjsewell** [comment]
for the PR, assuming that works, you can also remove: https://github.com/aiidateam/aiida-core/blob/bb2098e7035eda1526960897dec02fbd0a1a6a14/docs/source/conf.py#L24-L25

**Comment by chrisjsewell** [comment]
> Removing `formats` shouldn't be necessary - it's the default value anyway, but since there's a long comment explaining why it's empty maybe explicit is better here.

fair

**Comment by chrisjsewell** [comment]
> It also ignored the fail_on_warning setting, so I think it is not even reading the yaml or it is failing to parse it, but I cannot see a message of this anywhere

and that was because you put it under `mkdocs` not `sphinx` btw

**Comment by sphuber** [comment]
> and that was because you put it under `mkdocs` not `sphinx` btw

Yeah, just noticed that and push a new commit. Still, doesn't explain why the install command didn't get triggered :confused:

**Comment by greschd** [comment]
I'd guess it interpreted it as an old-version file and is just lenient in parsing..

**Comment by chrisjsewell** [comment]
you still haven't marked it as `version: 2` yet 😬

**Comment by sphuber** [comment]
Ah, didn't know that was required. Strange though, they say the following:
> Read the Docs validates every configuration file. Any configuration option that isn’t supported will make the build fail. 

So I would have thought the build to fail then...

**Comment by greschd** [comment]
> 
> > Read the Docs validates every configuration file. Any configuration option that isn’t supported will make the build fail.
> 
> So I would have thought the build to fail then...

Only says that on the `v2` page though 🤣

**Comment by chrisjsewell** [comment]
well at least it fails now: https://readthedocs.org/projects/aiida-core/builds/11337380/
baby steps 😆

**Comment by greschd** [comment]
Well, yeah.. that's because `pytest` is actually a proper dependency of AiiDA (for the fixture features), but we only define it in the `testing` extra.

**Comment by chrisjsewell** [comment]
opened a whole can of worms here 👀

**Comment by greschd** [comment]
Yeah, IMO everything that's being imported somewhere in `aiida/*` should be a "proper" dependency. Of course if we enable the nitpicking then it might just be we have to install with `.[docs,atomic-tools,rest,<...>]` on RTD.

**Comment by chrisjsewell** [comment]
no just `.[docs,tests]` is fine: https://github.com/aiidateam/aiida-core/blob/bb2098e7035eda1526960897dec02fbd0a1a6a14/.circleci/config.yml#L18

**Comment by greschd** [comment]
Interesting.. do you know why it needs `pytest` but not the other optional dependencies? Does it mean the `atomic-tools` extras don't get documented, for example?

**Comment by chrisjsewell** [comment]
well there just not imported during module imports, which I guess is the only thing that autodocs triggers (and maybe  classmethods?)

**Comment by chrisjsewell** [comment]
for example: https://github.com/aiidateam/aiida-core/blob/bb2098e7035eda1526960897dec02fbd0a1a6a14/aiida/orm/nodes/data/structure.py#L91-L110

in this case then they are IMO legitimate extras

**Comment by chrisjsewell** [comment]
yay https://readthedocs.org/projects/aiida-core/builds/11337534/ thanks @sphuber

**Comment by greschd** [comment]
Hmm, but what about https://github.com/aiidateam/aiida-core/blob/bb2098e7035eda1526960897dec02fbd0a1a6a14/aiida/restapi/api.py#L17-L19

.. do we just need to add an intersphinx entry for `pytest`?

Still think we should make "proper" extras for the dependencies of the fixtures and `aiida_sphinxext`.

**Comment by chrisjsewell** [comment]
magic 🤷 

well lets leave that for another issue, I just want to close [this PR](https://github.com/aiidateam/aiida-tutorials/pull/228) that I thought was just going to be a quick thing at the start of the day lol

**Comment by chrisjsewell** [comment]
btw @greschd I believe I am still waiting on a presentation from you 😜 ?

**Comment by greschd** [comment]
yeah.. hope I get around to doing it next week 😬

**Comment by chrisjsewell** [comment]
indeed, no more procrastinating around github issues until its done! 
Preferably early next week, given the tutorial is the week after ta

**Comment by ltalirz** [comment]
I've noticed the issue with the entry points with aiida-diff as wee https://readthedocs.org/projects/aiida-diff/builds/11309514/  which uses the `.readthedocs.yaml` https://github.com/aiidateam/aiida-diff/blob/master/.readthedocs.yml

So, to summarize, what is the currently suggested course of action?

**Comment by chrisjsewell** [comment]
> So, to summarize, what is the currently suggested course of action?

`version: 2`

**Comment by chrisjsewell** [comment]
and soon there should be nothing stopping us moving to directly using RTD as the CI runner: https://github.com/readthedocs/readthedocs.org/issues/7240#issuecomment-650386632 😄

**Comment by ltalirz** [comment]
> version: 2

Sorry, it seems I'm missing something - this is now my .readthedocs.yaml
https://github.com/aiidateam/aiida-diff/blob/fix-docs/.readthedocs.yml

and I added the reentry scan in the conf.py
https://github.com/aiidateam/aiida-diff/blob/eff6406b541e06795b93f5bc46f68a9d079eaaa6/docs/source/conf.py#L24-L25:

but my build still fails as before:
https://readthedocs.org/projects/aiida-diff/builds/11339478/

What am I doing wrong?

**Comment by ltalirz** [comment]
About the original issue I'm wondering: should we add the `reentry scan` to `load_documentation_profile`?

Granted, it would be a be a bit of a side effect, but effectively that is what people are going to need when they load a documentation dummy profile...

By the way, I think a slightly more "proper" way would be:
```
from aiida.plugins.entry_point import ENTRYPOINT_MANAGER
ENTRYPOINT_MANAGER.scan()
```
which (in principle) should work independently of the entry point manager used by AiidA - I discovered some time ago that `pkg_resources` actually doesn't have a `scan` method, so I'm not quite sure whether this is something that changed or was just overlooked.

**Comment by chrisjsewell** [comment]
> What am I doing wrong?

Hmm, yeh I can't see any obvious errors. The only difference I can see is you could try adding:
`sys.path.append(os.path.join(os.path.split(__file__)[0], os.pardir, os.pardir))`
I wouldn't of thought this is needed any more, but maybe it is doing something?

Also maybe try importing `aiida_diff.data` in the `conf.py` and see if it throws the same error

**Comment by chrisjsewell** [comment]
another thing I note is that you are installing sphinx version 3, whereas aiida-core restricts to <3. Maybe try constraining sphinx to < 3 in setup.json

**Comment by sphuber** [comment]
> and I added the reentry scan in the conf.py
> https://github.com/aiidateam/aiida-diff/blob/eff6406b541e06795b93f5bc46f68a9d079eaaa6/docs/source/conf.py#L24-L25:

This should not be necessary. Just installing the package should be enough, which you are doing now.

> but my build still fails as before:
> https://readthedocs.org/projects/aiida-diff/builds/11339478/
> What am I doing wrong?

The problem in your case doesn't seem to be with the entry points not being found but rather a straight up import error:
```
ModuleNotFoundError: No module named 'aiida_diff.data'
```
It seems that it cannot import this module and that is why it cannot load the entry point. It is a bit weird, because it seems to install the `aiida-diff` package. Maybe as Chris says, the seemingly pointless action of directly adding the repo directory to the path in the `conf.py` is necessary?

**Comment by greschd** [comment]
@ltalirz I think you're missing ``aiida-diff.data`` in the ``packages`` of ``setup.py``. See [this build](https://readthedocs.org/projects/greschd-aiida-diff-test/builds/11341641/) on [this branch](https://github.com/greschd/aiida-diff/blob/fix-docs/setup.py#L10).

**Comment by ltalirz** [comment]
@greschd Thanks a lot for looking into this and finding a fix!

I always assumed `find_packages()` was just for discovering all the top-level packages in the repo - I was not aware before that one needs to specify subpackages as well.

**Comment by greschd** [comment]
Yeah, it's not too intuitive.. 

Actually, looking at this in `aiida-core`, I think we need to add an `include` or `exclude` clause.. right now, it seems `tests`, `docs` and `utils` are also included.

Probably the best would be ``include=['aiida', 'aiida.*']``.

**Comment by ltalirz** [comment]
Very good point @greschd ; I've opened https://github.com/aiidateam/aiida-core/issues/4205 for this and will fix it asap
